dask/array/random.py

```diff
@@ -69,10 +109,15 @@ def _wrap(self, func, *args, **kwargs):
         name = 'da.random.{0}-{1}'.format(func.__name__, token)
         keys = product([name], *([range(len(bd)) for bd in chunks]
                                + [[0]] * len(extra_chunks)))
-        vals = ((_apply_random, func.__name__, seed, size, args, kwargs)
-                for seed, size in zip(seeds, sizes))
-        dsk = dict(zip(keys, vals))
 
+        vals = []
+        for seed, size, slc in zip(seeds, sizes, slices):
+            arg = [ar[slc] if isinstance(ar, (np.ndarray, Array)) else ar for ar
+                   in new_args]
```

Comment:
OK, two issues here, one for numpy arrays (which isn't very important) and one for dask.arrays (which is fairly important)

### NumPy (can be skipped)

It would be slightly nicer in a distributed computing setting if we didn't put numpy arrays into tasks, but instead added them directly to the graph

```python
>>> my_numpy_array = np.random.random(...)

>>> dsk = {'x': (add, my_numpy_array, 1), 'y': (sub, my_numpy_array, 2)}  # somewhat inefficient

>>> name = 'array-' + tokenize(my_numpy_array)
>>> dsk = {name: my_numpy_array, 'x': (add, name, 1), 'y': sub, name, 2)}  # more efficient
```

This allows the schedulers to track potentially larger datasets.  This is only important in a distributed setting though and only if users put in large numpy arrays though.  We could just leave this as future work.

### Dask

We don't want to put dask collections within dask graphs.  Instead for each output block we want to find the appropriate key from the input dask array to insert here.  In other words we don't want to put in `x`, we want to put in `(x.name, i, j, k)` based on the output index of this block.  Otherwise we're going to start calling dask.compute within dask.compute calls, which gets ugly pretty fast and results in a lot of unnecessary computation.

Generally for broadcasting operations on dask.array one can use the function `atop` to manage all of this.  It gets pretty tricky pretty fast.

Alternatively, you could punt on supporting dask.arrays in this and only support numpy arrays.